### PR TITLE
Memory-efficient push via pebble

### PR DESCRIPTION
**Pull request by Vultaire**
Uses a generator to efficiently send file data in chunks to pebble.

To view the diff of just the new parts: https://github.com/Vultaire/operator/compare/harness-pebble-fs...issues/576_push

### CODE REVIEW

File: test/test_testing.py [hunk 1]
```diff
@@ -2751,6 +2751,19 @@ def _test_push_and_pull_data(self, original_data, encoding, stream_class):
             received_data = infile.read()
         self.assertEqual(original_data, received_data)
 
+    def test_push_and_pull_larger_file(self):
+        # Intent: to ensure things work appropriately with larger files.
+        # Larger files may be sent/received in multiple chunks; this should help for
+        # checking that such logic is correct.
+        data_size = 2**20
+        original_data = "".join(chr(i % 128) for i in range(data_size)).encode('ascii')
```
Comment (benhoyt) [inline test/test_testing.py:2759]: Would `1024*1024` be a clearer way to specify 1MB than `2**20`? Would `os.urandom(data_size)` work just as well for this use case (and be simpler and more efficient) than the join/range loop?

File: ops/pebble.py [hunk 2]
```diff
@@ -1083,6 +1083,7 @@ def __init__(self, socket_path=None, opener=None, base_url='http://localhost', t
         self.opener = opener
         self.base_url = base_url
         self.timeout = timeout
+        self.block_size = 102400  # hard-coded for now
```
Comment (benhoyt) [inline ops/pebble.py]: What about writing this as `100*1024` for clarity?

File: ops/pebble.py [hunk 3]
```diff
@@ -1579,28 +1573,47 @@ def _make_auth_dict(permissions, user_id, user, group_id, group) -> typing.Dict:
...
-            b'Content-Disposition: form-data; name="files"; filename="', path_escaped, b'"\r\n',
-            b'\r\n',
-            content, b'\r\n',
-            b'--', boundary, b'--\r\n',
-        ])
         content_type = 'multipart/form-data; boundary="' + boundary.decode('utf-8') + '"'
-        return b''.join(parts), content_type
+
+        def generator():
+            nonlocal source
```
Comment (benhoyt) [inline ops/pebble.py:1591]: Nice use of a generator! I didn't know urllib.Request could take a generator -- that's great.

Side note: `source` doesn't need to be a `nonlocal` -- you only need `nonlocal` if you're assigning to a variable (`nonlocal` and `global` have to do with where a name is bound / assigned). Just referencing it is fine without any `nonlocal`.

File: ops/pebble.py [hunk 4]
```diff
@@ -1069,6 +1069,8 @@ def read1(self, n=-1):
 class Client:
     """Pebble API client."""
 
+    _chunk_size = 8192
```
Comment (benhoyt) [inline ops/pebble.py:1072]: Why is one of the chunk sizes 8KB and the other 100KB? (block_size)

File: test/test_testing.py [hunk 5]
```diff
@@ -2751,6 +2751,19 @@ def _test_push_and_pull_data(self, original_data, encoding, stream_class):
             received_data = infile.read()
         self.assertEqual(original_data, received_data)
 
+    def test_push_and_pull_larger_file(self):
```
Comment (benhoyt) [inline test/test_testing.py:2754]: Nice. Would be great if while you're at it you can add a couple of tests to `TestRealPebble` so push and pull can be easily tested against the real Pebble server too.

**Review by benhoyt — APPROVED** [review]
Couple of small points, but looks good to me, thanks!

File: ops/pebble.py (hunk 3 above)
Comment (Vultaire) [reply ops/pebble.py:1591]: I think I had assigned to source in a previous version while developing, added the keyword out of necessity at the time, then didn't check if it was still needed.  I can remove it; it looks like it's no longer needed.

File: ops/pebble.py (hunk 4 above)
Comment (Vultaire) [reply ops/pebble.py:1072]: If I'm remembering this right (couple of MRs here, this applies to at least one of them) - I think I totally forgot about the 100KB chunk size (which is what I added during initial implementation), and the 8192 was a chunk size I saw already in use elsewhere in the code that I tried to centralize for reuse later.

I'll review and likely drop the 100KB one.

File: test/test_testing.py (hunk 5 above)
Comment (Vultaire) [reply test/test_testing.py:2754]: So, here's an interesting bit (and the reason why this MR includes a bunch of seemingly unrelated stuff from #645): this test was added to a test mixin which is used for testing both against the mock Pebble backend and the real Pebble.

Actually those tests in #645 are something that's worth perhaps some extra thought.  They're in the test_testing module, but they also test real pebble.  I kind of feel like the overall "domain" is pebble, and they should be moved to test_pebble.py.

File: ops/pebble.py (hunk 2 above)
Comment (Vultaire) [reply ops/pebble.py]: Or replacing w/ the 8192 value used elsewhere...  I'll look into this.

File: test/test_testing.py (hunk 1 above)
Comment (Vultaire) [reply test/test_testing.py:2759]: 1024*1024: sure
urandom: thanks for the hint here, I like the idea

File: ops/pebble.py (hunk 3 above)
Comment (petevg) [reply ops/pebble.py:1591]: I like the generator, too. Nicely Pythonic, and an elegant way of doing things.

**Review by petevg — APPROVED** [review]
Approve of (what I believe) are the focused changes in this PR. Sweet little generator, simple handing of bytes vs. strings, and a bunch of tests (which is good).

I think that I'd like to hammer on this more in integration tests, to catch any edge cases that my brain isn't catching. But I am +1 to what is written here.

**Comment by Vultaire** [comment]
@petevg For transparency, keep in mind that most the tests here are simply from #645.  I only added one extra test here, for push/pull of a larger file which would trigger chunking.  However, in the push case I feel like that may be sufficient as the other tests do provide a significant harness for these changes.

Thanks for the review!

File: ops/pebble.py (hunk 2 above)
Comment (Vultaire) [reply ops/pebble.py]: Looks like I already addressed this in a later change.

**Comment by jnsgruk** [comment]
Despite failing tests, I think this is ready for broader review while the minor test issues are worked through.

**Comment by jameinel** [comment]
I found it a bit hard to do a review of just this PR since it depends on #645, if possible can we land that one first and then review this one?

**Review by jameinel — COMMENTED** [review]
(AFAICT the only actual change here above #645 is just to turn the core inner loop from returning a list to using a generator so we don't have to pull in all the bytes ahead of time, which sounds great, but hard to actually tease apart what is changing.)

**Comment by jnsgruk** [comment]
> (AFAICT the only actual change here above #645 is just to turn the core inner loop from returning a list to using a generator so we don't have to pull in all the bytes ahead of time, which sounds great, but hard to actually tease apart what is changing.)

Yep, I think you can see just the changes for this PR [here](https://github.com/Vultaire/operator/compare/harness-pebble-fs...issues/576_push)

**Review by jameinel — APPROVED** [review]
I don't see any user visible changes here, just a more efficient implementation if you are dealing with actual File objects.
Have we done direct testing of say a 1GB file and seen whether the charm and/or pebble are not actually wasting lots of memory?

**Review by petevg — COMMENTED** [review]
I am still +1 on this. 

But we should fix the failing test. (I think that the fault lies in the test suite, and it's going to be a fun one to mock out ...)

**Review by petevg — APPROVED** [review]
I'm happy to see how much simpler this PR gets w/ the base PR merged!

I am still very much on board with this generator. Elegant, readable, and simple. +1

**Comment by petevg** [comment]
I'm grabbing this to look into the failing tests.

The root cause seems to be that the mocked out object we're using in our unit tests doesn't have a .decode property, though it might go deeper than that. (IIRC, there was more convo around this earlier in the PR, or in another PR ...)

**Comment by jnsgruk** [comment]
Thanks @petevg!

**Comment by benhoyt** [comment]
Nice diabolical PR number, by the way! :-)
